File: src/main/kotlin/com/intellij/ml/llm/template/utils/MethodSignature.kt
package com.intellij.ml.llm.template.utils

import com.github.javaparser.ast.body.CallableDeclaration
import com.github.javaparser.ast.type.Type
import java.util.regex.Pattern

data class Parameter(val name: String, val type: String)

data class MethodSignature(val methodName: String, val paramsList: List<Parameter>, val returnType:String, val modifier: String){
    companion object{
        fun getMethodSignatureParts(methodSignature: String): MethodSignature? {
            val methodSignatureRegex = Pattern.compile("(\\w+) (\\w+)(\\(.*\\)) : ((\\w(<\\w+>)?)+)")
            var m = methodSignatureRegex.matcher(methodSignature)
            if (m.find()) {
//            val methodInformation = methodSignature.strip().split(" ")
                val modifier = if (m.group(1)=="package") "packageLocal" else m.group(1)
//            val methodNameAndParams = methodInformation[1]
                val methodName = m.group(2)
                val methodParams = m.group(3)
                val returnType = m.group(4)
                return MethodSignature(methodName, getParamsList(methodParams), returnType, modifier)
            }
            return null
        }

        fun getParamsList(methodParams: String): List<Parameter>{
            if (methodParams=="()")
                return emptyList()
            return methodParams.substring(1, methodParams.length-1).split(", ")
                .map {
                    val nameAndType = it.strip().split(" ")
                    Parameter(nameAndType[0], nameAndType[1])
                }
        }
    }

    fun compare(signature: CallableDeclaration.Signature): Boolean{
        if (this.methodName != signature.name) return false
        if (signature.parameterTypes.isEmpty()) return this.paramsList.isEmpty()
        return signature.parameterTypes

                    .mapIndexed { index: Int, type: Type -> this.paramsList.getOrNull(index)?.type == type.asString() }
                    .reduce { acc, parameter -> acc && parameter  }
    }

}

File: src/main/python/mm_analyser/refactoring_miner_processing/MethodSignature.py
import re

class Parameter:
    def __init__(self, param_type, param_name):
        self.param_type = param_type
        self.param_name = param_name

class MethodSignature:
    def __init__(self,
                 original_signature: str,
                 method_name: str,
                 params: list[Parameter],
                 return_type: str,
                 modifier: str):
        self.original_signature = original_signature
        self.method_name = method_name
        self.params = params
        self.return_type = return_type
        self.modifier = modifier

    def __str__(self):
        return self.original_signature

    @staticmethod
    def get_method_signature_parts(method_signature: str):
        method_signature_regex = re.compile(r"(\w+) (\w+)(\(.*\)) : ((\w(<\w+>)?)+)")
        m = method_signature_regex.search(method_signature)

        if m:
            modifier = "packageLocal" if m.group(1) == "package" else m.group(1)
            method_name = m.group(2)
            method_params = m.group(3)
            return_type = m.group(4)
            return MethodSignature(
                method_signature,
                method_name,
                MethodSignature.get_params_list(method_params),
                return_type,
                modifier)

        return None

    @staticmethod
    def get_params_list(method_params: str):
        if method_params == "()":
            return []

        params_list = method_params[1:-1].split(", ")

        return [Parameter(param.split(" ")[1], param.split(" ")[0])
                for param in params_list]



